[1 - 3단계 방탈출 사용자 예약] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 이스트! 이번 미션의 리뷰이 도기 입니다!
리뷰어로 만나게 되어서 영광입니다🙇‍♂️
부족한 코드지만 리뷰 잘 부탁 드립니다 : )

이번 미션 시작 커밋 입니다! [새로운 변경 사항 커밋 지점](https://github.com/woowacourse/spring-roomescape-member/pull/240/files/9fe34ab75be32d3b0a1240fa44522b86ad49c477..2249d956e6f53c68cf49d7054d97046e0fc0cb06)

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
  - [셀프체크](https://prolog.techcourse.co.kr/studylogs/4314)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

기존의 테스트 코드를 수정하지 못해 주석처리 한 상태입니다. 그래서 기본을 못 지킨 것 같습니다!
이외에도 메서드 라인 수, 클래스 인스턴스 개수, 매직넘버 상수화 등을 지키지 못한 것 같습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

기능 확장과 테이블 추가로 인해 엔티티 간 연관관계가 복잡해졌습니다.
그에 따라, 각 기능에 대한 책임이 어떤 객체(또는 계층)에 있어야 하는지 판단이 어려워졌습니다.

이번 코드 리뷰에서는 전체적인 레이어 구조(Service, Domain, Repository) 가 적절하게 분리되어 있는지, 각 클래스가 올바른 책임을 가지고 있는지를 중점적으로 봐주셨으면 합니다.

또한, 제가 미처 고려하지 못했거나 추가로 학습하면 좋을 개념이나 키워드가 있다면, 피드백 해주시면 좋을 것 같습니다!
이외에 특정 코드부분에 대한 질문은 코멘트로 남기겠습니다 : )


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

CodingMasterLSW committed:
[1-3단계 - 방탈출 예약 관리] 젠슨(이성우) 미션 제출합니다.  (#216)

* docs: 초기 기능명세서 작성

* feat: admin 페이지 응답 구현

* feat: Person 구현

* feat: ReservationTime 구현

* feat: Reservation 및 Reservation 일급컬렉션 구현

* refactor: 오타 수정 (nane -> name)

* refactor: Reservations 생성 방식 변경 및 Reservation id 타입 수정(int -> Long)

* feat: 예약 목록 조회 기능 구현 및 테스트 코드 작성

* refactor: Person id 자료형 변경 (int -> long)

* rename: Dto 이름 변경 (ReservationDto -> ReservationResponseDto)

* feat: 예약 추가 기능 구현

* feat: 예약 삭제 기능 구현

* fix: 예약 시간 표기 오류 수정 (format 형식 변경으로 해결)

* refactor: 일부 클래스 Record 적용 및 디미터 법칙 적용

* rename: RoomEscape -> Roomescape 통일

* refactor: Controller 기능 별 분리

* rename: RoomescapeController -> ReservationController

* test: 일단계() api 주소 변경 (/ -> /admin)

* refactor: 각각의 Index 위치 변경 (controller -> domain)

* rename: AdminController -> ViewMappingController

* test: Reservations 메서드 테스트 코드 작성 (save, delete)

* refactor: index값 공유 문제로 인해 Controller에서의 주입으로 변경

CodingMasterLSW committed:
[4-9단계 - 방탈출 예약 관리] 젠슨(이성우) 미션 제출합니다. (#281)

* feat: DB 테이블 초기화 설정

* feat: DB에 있는 예약 기록 조회

* feat: DB 예약 기록 CREATE, DELETE 구현

* docs: 시간 관련 명세서 추가

* rename: ReservationTime -> ReservationDateTime

* feat: 전체 예약가능 시간 조회 API 구현

* feat: 특정 시간 삭제 기능 구현

* refactor: reservationTime api 주소 변경 (/admin/times -> /times)

* refactor: 테이블 스키마 재정의로 인한 로직 수정

예약 수정, 조회 쿼리 수정 및 시간 추가, 조회 쿼리 수정을 진행하였음.

* remove: 레거시 코드 삭제

* test: 테이블 스키마 재정의 단계 테스트 코드 작성

* refactor: 레이어드 아키텍처 적용

* refactor: 조회 속도 향상을 위한 캐싱 처리

* refactor: ReservationDao 인터페이스화 및 ReservationService 테스트코드 작성

* refactor: ReservationTimeDao 인터페이스화 및 ReservationTimeService 테스트코드 작성

* remove: 레거시 코드 삭제

* remove: 레거시 코드 삭제

* refactor: 생성자 체이닝 방식 적용

* test: reservations 테스트코드 작성

* refactor: 동시성 문제 해결을 위한 synchronized 적용

* refactor: 캐시 객체 삭제

* refactor: person record -> class 변경

단순한 데이터 운반책으로 설계한 객체가 아니기에 변경을 진행하였음

* feat: Reservation, ReservationTime equals & hashcode 재정의 및 reservationDao 테스트코드 작성

* test: ReservationTimeDao 테스트코드 작성 및 given, when, then 적용

solarsalt committed:
docs: add PR template

CodingMasterLSW committed:
docs: 초기 기능명세서 작성

CodingMasterLSW committed:
test: pr 테스트용 커밋

CodingMasterLSW committed:
test: 오류 나는 테스트 코드 주석처리

CodingMasterLSW committed:
feat: 시간 예약 페이지 매핑 구현

CodingMasterLSW committed:
refactor: 예약 시간 관련 API 응답 상태코드 변경

- 저장: 200 -> 201
- 삭제: 200 -> 204

CodingMasterLSW committed:
refactor: 예약 관련 API 응답 상태코드 변경

- 저장: 200 -> 201
- 삭제: 200 -> 204

CodingMasterLSW committed:
feat: ReservationRequestDto 검증 로직 추가

CodingMasterLSW committed:
feat: ReservationTimeRequestDto 검증 로직 추가

CodingMasterLSW committed:
feat: 지나간 날짜에 대한 예약 예외처리

- Date를 ResevationDate로 포장하였음.

CodingMasterLSW committed:
feat: 현재 날짜로 예약을 하는 경우, 과거 시간으로 예약 생성 불가능하도록 예외 처리

- 도메인 생성자 -> Service로 예외 처리 로직을 전면 수정하였음

CodingMasterLSW committed:
feat: 이미 존재하는 시간에 대한 예약이 존재할 때 삭제 불가 예외 추가

CodingMasterLSW committed:
feat: 동일한 날짜와 시간을 예약하는 경우 예외 처리 추가

CodingMasterLSW committed:
feat: 예외에 따른 상태코드 반환 기능 추가

- GlobalExceptionHandler 도입

CodingMasterLSW committed:
docs: 2단계 기능요구사항 정리

CodingMasterLSW committed:
feat: 테마 관리 페이지 매핑 추가

CodingMasterLSW committed:
feat: 테마 테이블 추가

- 테마 도메인 객체 생성

CodingMasterLSW committed:
feat: 테마 조회 API 구현

CodingMasterLSW committed:
feat: 테마 추가 API 구현

CodingMasterLSW committed:
feat: 테마 삭제 API 구현

CodingMasterLSW committed:
refactor: 테마 API가 구체적인 상태코드를 반환하도록 수정

CodingMasterLSW committed:
refactor: 예약 조회에서 테마도 포함해 반환하도록 수정

CodingMasterLSW committed:
docs: 3단계 초기 기능명세서 작성

CodingMasterLSW committed:
test: 테스트 환경을 위해 기존 테스트 임시 주석처리

CodingMasterLSW committed:
feat: 유저 예약 페이지 매핑 구현

CodingMasterLSW committed:
feat: 예약 가능한 테마 렌더링 구현

CodingMasterLSW committed:
feat: 테마를 선택하면, 예약 가능한 시간을 보여줄 수 있다.

CodingMasterLSW committed:
refactor: 날짜별, 테마별 예약 현황을 확인할 수 있도록 수정

- 이미 예약이 되었다면 선택하지 못함

CodingMasterLSW committed:
chore: 더미데이터 추가

CodingMasterLSW committed:
feat: 일주일 기준으로 예약이 많은 테마 10개를 확인하는 API 추가

CodingMasterLSW committed:
refactor: 테스트 코드 오류로 인한 더미데이터 삭제

dye0p commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,84 @@
+        this(null, person, date, reservationTime, theme);
+    }
+
+    public void validateDateTime(
+        ReservationDate date,
+        ReservationTime reservationTime,
+        LocalDateTime currentDateTime) {
+        if (date.isCurrentDay(currentDateTime.toLocalDate()) &&
+            reservationTime.isBefore(currentDateTime.toLocalTime())) {
+            throw new IllegalArgumentException("과거 시간으로는 예약할 수 없습니다.");
+        }
+    }
```
기존에는 `Reservation` 객체의 유효성 검증`validateDateTime()`을 생성자 내부에서 수행하여,
객체가 생성되는 시점에 유효한 상태를 보장하고, 객체의 응집도를 유지하려고 했습니다.

하지만, `RowMapper`를 통한 `findAll()` 조회 시 이미 과거의 예약 데이터가 포함된 경우,
생성자 내부의 검증 로직으로 인해 예외가 발생하며 정상적인 조회가 불가능한 문제가 발생했습니다.

이에 따라 유효성 검증을 생성자 밖으로 분리하고, `Service` 계층에서 명시적으로 호출하는 방식으로 구조를 변경했습니다.

`ReservationService.class`
``` java
    public ReservationResponseDto saveReservation(ReservationRequestDto reservationRequestDto) {
        LocalDateTime currentDateTime = LocalDateTime.now();
        ...
        Reservation reservation = new Reservation(person, date, reservationTime, theme);
        reservation.validateDateTime(date, reservationTime, currentDateTime);
```
각각의 방식에는 장단점이 있다고 생각합니다. 

### 생성자 방식

장점
- 생성하는 시점에 유요한 상태를 보장할 수 있음
- 도메인의 책임이 분산되지 않고, 도메인 내부에서 처리하여 응집도가 높아짐

단점
- RowMapper로 객체를 복원할 때 과거의 경우가 있다면 객체 생성이 실패하여 조회 불가

### 외부에서 호출 (service 계층)

장점
- 조회시 예외없이 조회가능

단점
- 외부에서 호출하기 때문에 호출이 누락될 수 있음(개발자의 실수 등)
- 도메인의 생성과 관련된 로직이 외부로 노출되어 응집도가 낮아짐
- 생성시 검증을 하지 않기 때문에 객체의 불완전한 상태

---

객체를 생성하는 시점에 유효한 상태를 보장해야 한다는 관점에서 보면, 생성자에서 검증하는 것이 자연스럽다고 생각하는데, 조회시 예외처리가 되어 현실적으로 힘들다고 생각합니다. 그렇다면 지금처럼 `Service`에서 별도로 검증을 호출하는 방식은 어떻게 생각하시나요?

이런 구조는 도메인 객체와 엔티티 객체를 분리해야 한다는 신호로 볼 수 있을까요? 아니면 현실적인 타협으로 받아들여도 무방할까요? 이스트의 생각이 궁금합니다!

EastHShin replied:
예약이라는 도메인 객체의 생성 조건을 다시 한번 점검해보면 좋을 것 같아요.
과거의 예약이 존재한다 해서 reservation 객체를 생성시 예외가 발생해야 할까요? 과거 예약도 예약이기 때문에 생성은 할 수 있어야 한다고 생각해요. 생성한 후에, 그 다음 조건으로 시간을 검증하는 메서드를 호출해주면 될 것 같습니다 :)
도메인의 완전한 상태라는 것의 정의를 다시 한번 생각해보시면 받아들이기 쉬울 것 같아요!

dye0p replied:
남겨주신 피드백으로 다시 생각해본 결과, `과거의 예약도 예약이다`에 동의할 수 있을 것 같아요. 
`과거의 예약이라고 예약을 생성할 수 없다`는 것은 예약 객체의 유효성을 판단하는 비즈니스 규칙이지, 객체 생성 자체를 막아야 하는 이유는 아니라고 판단했습니다. 따라서 지금과 같은 구조를 유지하는것이 적절하다는 결론을 내렸습니다!

dye0p commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,96 @@
+    private void validateAlreadyExistDateTime(ReservationRequestDto reservationRequestDto,
+        ReservationDate date) {
+        if (reservationDao.findByDateAndTime(date, reservationRequestDto.timeId()) != 0) {
+            throw new InvalidReservationException("중복된 날짜와 시간을 예약할 수 없습니다.");
```
`ReservationService`에서 중복된 날짜와 시간으로 예약을 시도할 경우
현재는 `InvalidReservationException`을 통해 `500 Internal Server Error`를 반환하고 있습니다.

이렇게 처리한 이유는, 사용자 입장에서 이 요청은 겉보기에 잘못된 입력은 아니므로  
서버 쪽 문제로 해석할 수 있다고 생각했기 때문입니다.

하지만 다시 생각해보면, 중복 예약은 도메인 규칙 위반에 해당하는 **비즈니스 예외**이고,  
이는 결국 클라이언트가 유효하지 않은 요청을 보낸 것이라고 볼 수 있을 것 같다는 생각이 듭니다.

이 경우 `400 Bad Request` 혹은 `500 internal server error` 중 어떤 상태코드가 더 적절할지 고민이 됩니다.  
중복 예약 같은 비즈니스 예외도 `4xx` 범주로 보는 것이 일반적인지,  
혹은 서버 오류로 간주해 `500`으로 처리하는 것이 괜찮은 선택인지 궁금합니다.

상태 코드에 대한 이스트의 생각은 어떠신가요?

EastHShin replied:
중복예약 시도는 결국 클라이언트 측에서 그러한 시도를 한 것이기 때문에, 400 상태가 더 알맞다고 생각 되네요ㅎㅎ

dye0p replied:
400 상태코드를 반환하도록 수정하였습니다! 

[refactor: InvalidReservationException 예외 발생시 응답 상태코드 변경](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/ea575ed840919d7bba01b60cb28121b6785f1207)

EastHShin commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,49 @@
+        return reservationService.getAllReservations();
+    }
+
+    @GetMapping("/reservations/dates/{date}/themes/{themeId}/times")
```
모든 것을 http get 메서드 uri 를 통해 표현할 필요는 없습니다. 우리는 예약 가능한 시간을 얻어오는 것이고, 이에 필요한 리소스는 http post 의 body 를 통해 표현하는 것이 확장하는데 있어서도 효율적일거에요 :)

dye0p replied:
기존 방식의 요청 URL에 대해서 고민이 있었는데, 말씀 해주신 방법도 있다는 걸 알게되었네요! 

적용을 해봤을 때 느낄 수 있었던 장점을 정리해보았습니다.
- URL의 간소화로 요구사항이 추가 되었을 때 URL 수정이 일어나지 않는다.
- 요청 세부사항을 외부로 노출시키지 않을 수 있다.

조금 더 유연한 사고를 할 수 있었습니다! 감사합니다 : )

[refactor: 예약 가능한 시간 요청 http 메서드 방식 변경(GET -> POST)](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/4d695f3a46a99fe2d6e318e51d8cf136f3104028)

EastHShin commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,49 @@
+        return reservationService.getAllBookedReservationTimes(date, themeId);
+    }
+
+    @PostMapping("/reservations")
```
`/reservations` 가 공통적으로 나타나는데, `@RequestMapping` 을 클래스 부분에 선언해 중복을 줄여볼까요?

dye0p replied:
적용하였습니다! 👍

[refactor:](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/587a61ca23945cc9bea8d6fb7ed2cfce2c0b555c) `@RequestMapping`[을 사용하여 요청 URL의 중복 제거](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/587a61ca23945cc9bea8d6fb7ed2cfce2c0b555c)

EastHShin commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,49 @@
+
+    @PostMapping("/reservations")
+    @ResponseStatus(HttpStatus.CREATED)
+    public ReservationResponseDto createReservation(
```
`ResponseEntity` 를 사용하지 않고 `@ResponseStatus` 를 사용해주신 이유는 무엇일까요? (바꾸라는 이야기는 아닙니다!)

dye0p replied:
이 부분은 페어와도 이야기한 부분이었습니다. 
그래서 `ReponseEntity` 와`@ReponseStatus`의 장단점을 비교해 보았습니다.

### ResponseEntity
장점
- 동적으로 상태코드를 응답할 수 있다. (분기 처리 가능)
- 상태코드와 `body`를 모두 조작할 수 있다.

단점
- 코드가 길어질 수 있다. (단점인지는 잘 모르겠습니다!)

### @ResponseStatus
장점
- 간결하게 어노테이션으로 상태코드를 응답할 수 있다.

단점
- 응답 body를 조작하기 어렵다 (`void`, `객체` 타입만 반환 가능)
- 동적으로 상태코드를 응답하기 어렵다. (메서드 레벨에 어노테이션으로 지정해주기 떄문)

지금의 단계에서는 둘 중 어떤 것 을 사용하던 상관이 없다고 생각했습니다. 그렇다면 간결하게 표현하고자 @ReponseStatus 를 사용하였습니다!

이스트는 어떻게 생각하시나요?

EastHShin replied:
ResponseEntity 를 굳이 사용할 필요없다면 사용해주신 방법도 좋은 것 같아요~
사용할 필요가 있을때 바꿔주는 방향으로 가면 좋겠네요!

EastHShin commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,20 @@
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(IllegalArgumentException.class)
```
IllegalArgumentException 자체를 예외 핸들링해 400 응답으로 내려주면, 자바 라이브러리 등에서 IllegalArgumentException 을
던지는 경우 예상치 못하게 400 응답으로 반환될 수도 있습니다. 어떻게 개선해보면 좋을까요?

dye0p replied:
> IllegalArgumentException 자체를 예외 핸들링해 400 응답으로 내려주면, 자바 라이브러리 등에서 IllegalArgumentException 을
던지는 경우 예상치 못하게 400 응답으로 반환될 수도 있습니다

이 부분을 생각하지 못했네요! 말씀하신대로 범용적인 예외를 전역에서 잡게되면 예상하지 못한 부분에서 정확한 예외가 발생하지 않을 것 같다고 생각합니다.

따라서 커스텀 예외를 생성하여 개선하였습니다!

[refactor: 클라이언트의 잘못된 요청에 대한 커스텀 예외 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/1f7c4c5271d597a46dab83c4926d33b4996d2e14)

EastHShin commented on src/main/java/roomescape/dto/ReservationTimeRequestDto.java:
```diff
@@ -0,0 +1,34 @@
+            throw new IllegalArgumentException("잘못된 시간입니다.");
+        }
+        validateParse(startAt);
+    }
```
spring valiadation 을 통해 컨트롤러에서 넘어오는 request body 데이터를 검증해볼까요?

dye0p replied:
적용하였습니다! 👍

[refactor: 요청 dto 검증 Spring validation 으로 전환](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/bf612a48d4a0c9393816519c844e8e42a28cb564)

EastHShin commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,42 @@
+import roomescape.service.ReservationTimeService;
+
+@RestController
+public class ReservationTimeController {
```
<img width="1111" alt="image" src="https://github.com/user-attachments/assets/0ea60fbc-fc40-47ac-aae4-cb0d2e967ba2" />

시간을 선택할 수 없군요ㅜㅜ 테마 API만 호출하고 있네요. 고쳐봅시다!

dye0p replied:
프론트엔드에서 기대하는 응답 값의 변수명과 dto의 변수명이 일치하지 않아서 생긴 문제였습니다! 수정하였습니다 👍

[fix: 프론트의 응답 필드명과 DTO 간 필드명 불일치 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/829fcb5c0308aac6f34207d2de0c9e07d898171e)

EastHShin reviewed (changes requested):
안녕하세요 도기~

이번 미션 함께 하게 된 이스트입니다!
질문에 대한 답변과 코멘트 몇 개 남겨 두었으니 확인 부탁드려요~

dye0p committed:
fix: 프론트의 응답 필드명과 DTO 간 필드명 불일치 수정

dye0p committed:
refactor: @RequestMapping을 사용하여 요청 URL의 중복 제거

dye0p committed:
refactor: 요청 dto 검증 Spring validation 으로 전환

- Spring validation 의존성 추가
- 기존의 dto 테스트는 실행할 수 없으므로 제거
- Controller 에서 사용가능 하도록 @Valid 어노테이션 추가

dye0p committed:
fix: 과거의 시간으로 예약하는 검증 로직 수정

dye0p committed:
refactor: 클라이언트의 잘못된 요청에 대한 커스텀 예외 추가

- 기존의 IllegalArgumentException 타입의 예외를 커스텀 예외로 변경

dye0p committed:
refactor: InvalidReservationException 예외 발생시 응답 상태코드 변경

- 500 INTERNAL_SERVER_ERROR -> 400 BAD_REQUEST

dye0p committed:
refactor: 예약 시간이 존재하는지 찾는 쿼리 변경

- 기존 예약시간이 존재하는 갯수를 반환하던 count 쿼리를 존재하는지 여부를 판단하는 exists 쿼리로 변경

- 존재하는지 여부만 판단하는 기능이라면 테이블을 모두 순회하며 count를 세는 것은 비효율적임

dye0p committed:
refactor: 코드 포멧팅 및 메서드 분리 리팩터링

dye0p committed:
refactor: 중복된 예약 검증 쿼리 변경(count -> exists)

dye0p committed:
feat: 이미 예약된 테마를 삭제하려는 경우 예외처리 추가

dye0p committed:
feat: 존재하지 않는 예약 시간 삭제 시도시 예외 처리 추가

dye0p committed:
test: 테스트 실패 해결

- ReservationTime 객체에 equals() & hashcode() 재정의

dye0p committed:
refactor: deleteTheme 메서드 내부 메서드 호출 순서 변경

- id에 해당하는 theme가 존재하는지 먼저 확인

dye0p committed:
refactor: 코드 포멧팅 및 메서드 배치 순서 리팩터링

dye0p committed:
fix: existsReservationBy 오타 수정

- 메서드 네이밍 오타 수정
- 쿼리문 문법 오타 수정

dye0p committed:
refactor: 예약 가능한 시간 요청 http 메서드 방식 변경(GET -> POST)

- 기존 PathVariable로 파라미터를 받았던 구조는 필요한 정보가 추가될 수록 수정이 번거롭고 계속해서 추가해주어야 하기 때문에 확장에 불리하다고 판단하였음.

- POST 요청을 통해서 요청 body에 필요한 요청 데이터를 추가하여 복잡한 URL을 단순화 하였음.

- 필요한 정보가 추가되어도 URL의 수정이 필요없음.

dye0p committed:
test: 주석 처리되었던 테스트 정상 동작하도록 수정

dye0p committed:
refactor: 테스트 코드 패키지 이동

dye0p committed:
refactor: findAllThemeOfRanks() 쿼리문 수정

- 최대 10까지 가져올 수 있도록 LIMIT 쿼리 추가
- 예약 수 별로 정렬 할 수 있도록 DESC 쿼리 추가

dye0p committed:
test: JdbcThemeDaoImpl 테스트 추가

dye0p committed:
remove: sql 더미 데이터 제거

dye0p committed:
fix: 같은 시간에 다른 테마 예약이 가능하도록 수정

- 기존 쿼리는 같은 날짜와 시간만으로 중복을 검사하고 있어, 같은 시간에 다른 테마를 예약하는 것이 불가능한 문제가 있었음.

- 예약을 생성할 때 themeId도 함께 검사하도록 원래 사용하던 메서드를 제거하고, existsReservationBy() 메서드를 사용하도록 변경하였음.

dye0p reviewed (commented):
안녕하세요 이스트! 말씀해주신 피드백 반영과 추가로 수정이 필요하다고 생각한 부분들을 수정하였습니다! 

리팩터링 중간중간 push를 해서 커밋이 코드레빗 코멘트와 섞여있어, 알아보기 힘들것 같아서 여기에 남겨놓겠습니다! (위에서 아래 순서 입니다.)

리뷰 잘 부탁 드립니다 : )

---

### 리팩터링 커밋 사항
[fix: 프론트의 응답 필드명과 DTO 간 필드명 불일치 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/829fcb5c0308aac6f34207d2de0c9e07d898171e)

[refactor:](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/587a61ca23945cc9bea8d6fb7ed2cfce2c0b555c) @RequestMapping[을 사용하여 요청 URL의 중복 제거](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/587a61ca23945cc9bea8d6fb7ed2cfce2c0b555c)

[refactor: 요청 dto 검증 Spring validation 으로 전환](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/bf612a48d4a0c9393816519c844e8e42a28cb564)

[fix: 과거의 시간으로 예약하는 검증 로직 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/0971cd96cb6da73a51cc711eefdb438fe146fc17)

[refactor: 클라이언트의 잘못된 요청에 대한 커스텀 예외 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/1f7c4c5271d597a46dab83c4926d33b4996d2e14)

[refactor: InvalidReservationException 예외 발생시 응답 상태코드 변경](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/ea575ed840919d7bba01b60cb28121b6785f1207) 

[refactor: 예약 시간이 존재하는지 찾는 쿼리 변경](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/3057c4c7e5e35519da362f291f71ae330449ea5a) 

[refactor: 코드 포멧팅 및 메서드 분리 리팩터링](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/55e798c37a66ad5f4b694988fcaf0c093d3bbaf1)

[refactor: 중복된 예약 검증 쿼리 변경(count -> exists)](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/31f9df3fa6243f31a6e42c19817a2baf41a444c4)

[feat: 이미 예약된 테마를 삭제하려는 경우 예외처리 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/00c312de99b1534d677846d50c221facaf67c428)

[feat: 존재하지 않는 예약 시간 삭제 시도시 예외 처리 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/4c57e563c6e10cb4a331b01f47d6cd4174c9a49e)

[test: 테스트 실패 해결](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/1af18abf92071d9287887b4e30a741b4982926a4) 

[refactor: deleteTheme 메서드 내부 메서드 호출 순서 변경](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/425c698b0cd97dc0b7d8a581bd4de8616bad0bf8) 

[refactor: 코드 포멧팅 및 메서드 배치 순서 리팩터링](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/c3c4d9c3221a0386ec46c0e1fd636ccbd47ba7fa)

[fix: existsReservationBy 오타 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/a3fa60db98584bed14ef132fb753c9df6b889a3d) 

[refactor: 예약 가능한 시간 요청 http 메서드 방식 변경(GET -> POST)](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/4d695f3a46a99fe2d6e318e51d8cf136f3104028) 

[test: 주석 처리되었던 테스트 정상 동작하도록 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/75213ab96541ae4ab8f5c7047d150b89b0c15f8e)

[refactor: 테스트 코드 패키지 이동](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/4c07f01ea18cc32bde26f7c73346b56522db3e8e)

[refactor: findAllThemeOfRanks() 쿼리문 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/1cea2b69e92009401dd9acd3d878442b3285afbb) 

[test: JdbcThemeDaoImpl 테스트 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/28f281a77546453c43da72ad5f57aa0eaf333c76)

[remove: sql 더미 데이터 제거](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/060783659e2476ab76e509dd3c76180ccfa5af01)

[fix: 같은 시간에 다른 테마 예약이 가능하도록 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/e5cdc9f41d5ca2ceef5a6ee90cf2c5d8808c8e23)

EastHShin commented on src/main/java/roomescape/dao/JdbcThemeDaoImpl.java:
```diff
@@ -0,0 +1,93 @@
+                        ) AS sub
+                        INNER JOIN theme ON sub.theme_id = theme.id
+                        ORDER BY sub.reservation_count DESC
+                        LIMIT 10
```
date 를 파라미터를 통해 동적으로 조정해준것 처럼, limit 도 그렇게 해주면 좋을 것 같아요~
일주일간의 랭킹 10위까지의 테마를 불러온다는 도메인 규칙이 sql 쿼리에 포함되어있는 것 보다는, 비즈니스 로직을 처리하는 부분에서 조절해주는 것이 응집성이 있기 때문입니다 :)

dye0p replied:
생각하지 못한 부분이네요! 말씀 해주신 것 처럼 최대 개수를 필터링 하는 조건도 비스니스적 요소로 봐야할 것 같습니다! 
추가로, 리팩터링 과정에서 고민이 생겼습니다.

**최대 조회 개수를 `Service` 내부에서 고정값으로 정의**할지, 
아니면 
**클라이언트 요청 파라미터로 받아서 처리**하는 게 더 나은지 고민이 됩니다.

### 🤔 클라이언트의 요청 책임일까? vs 클라이언트는 신경쓰지 말아야 할까?
개인적으로는 '**클라이언트가 꼭 최대 개수까지 신경 써서 요청해야 할까?**' 라는 생각이 들어서, Service 내부에서 제한하는 것이 더 적절하다고 판단하였습니다. 

하지만, 비즈니스 규칙이 자주 바뀔 가능성이나 유연성(20개를 보여줘야 하는 경우)을 고려하면 외부에서 받는 게 나을 것 같다는 생각도 듭니다.  

이에 대해서 이스트는 어떻게 생각하시나요?

[refactor: 인기테마 조회시 최대 개수를 외부에서 받을 수 있도록 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/5e5045ec2605c6c180d1172c8844112f81eacd1a)

EastHShin replied:
저는 이 경우에는 조회 개수에 대한 유연성을 주기 위해 API 요청을 통해 조절할 수 있도록 하는 편이에요~
다만, limit 개수가 무한정 늘어나는 경우 DB에 부하를 줄 수 있기 때문에, limit 의 최대 개수는 서버에서 조절해야 할 것 같네요 :)

EastHShin commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,26 @@
+        return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(InvalidReservationTimeException.class)
+    public ResponseEntity<String> handleInvalidReservationTime(InvalidReservationTimeException e) {
+        return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
+    }
```
작성해주신 커스텀예외가 아닌 예외가 발생했을때 공통적으로 처리해줄 수 있는 핸들러가 있으면 좋을 것 같아요~

dye0p replied:
커스텀 예외를 제외하고 다른 예외는 처리가 부족했네요! 이외의 예외를 처리할 수 있도록 리팩터링 했습니다!

[feat: 커스텀 예외 이외의 예외를 처리하는 예외 핸들러 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/64dc2abed8dbd80dd6efd58a79dea0116f600272)

EastHShin commented on src/main/java/roomescape/dao/JdbcThemeDaoImpl.java:
```diff
@@ -0,0 +1,93 @@
+    @Override
+    public void deleteTheme(Long id) {
+        String query = "delete from theme where id = ?";
+        jdbcTemplate.update(query, id);
```
존재하지 않는 리소스에 대해 삭제하려고 했을때는 삭제가 되지 않을텐데, 이에 대한 예외 처리를 해주면 어떨까요?

dye0p replied:
>아 이 부분에서 삭제 전 검증을 하고 있군요!
테마에 대한 데이터가 존재하는지 검증하는 로직을 서비스 계층애서 처리하지 않고, repository 계층에서 처리하는 건 어떤가요? 장단점을 비교해보시면 좋을 것 같네요!

해당 코멘트에서 해결하였습니다!
https://github.com/woowacourse/spring-roomescape-member/pull/240/files/e5cdc9f41d5ca2ceef5a6ee90cf2c5d8808c8e23#r2074659649

EastHShin commented on src/main/java/roomescape/service/ThemeService.java:
```diff
@@ -0,0 +1,58 @@
+    public void deleteTheme(Long id) {
+        validateNotFoundThemeBy(id);
+        validateExistsThemeBy(id);
+        themeDao.deleteTheme(id);
```
아 이 부분에서 삭제 전 검증을 하고 있군요!
테마에 대한 데이터가 존재하는지 검증하는 로직을 서비스 계층애서 처리하지 않고, repository 계층에서 처리하는 건 어떤가요? 장단점을 비교해보시면 좋을 것 같네요!

dye0p replied:
말씀 주신 피드백에 대해서 고민해보고 적용해보았습니다. 그리고 제가느낀 장단점을 정리해보았습니다!

### repository 계층에서 예외 처리

장점
- 해당 행이 존재하는지 여부는 db가 알고있다. 따라서 repository 계층에서 검증을 하는 것이 적절한 책임이다.
- 서비스 계층에서 번거로운 메서드 호출을 줄일 수 있다.

단점
- 더 많은 검증 사항이 생겼을 때 dao의 메서드 길이가 길어진다.
- 결국은 서비스로 책임을 분리해야 할 것 같다.

최종적으로는 변경을 하였습니다! 이유는 데이터의 존재여부는 db에서 판단하는것 것이 더 적절한 책임이라고 생각했기 때문입니다! 

이에 대해서 이스트는 어떻게 생각하시나요?

[refactor: 테마 삭제 유효성 검사 로직 repository 계층으로 이동](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/53ca3cdc0207ffd6a1c1f20b0293311c1aaff51d)
[refactor: 예약 시간 삭제시 검증 로직 repository 계층으로 이동](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/274f947dccce248a8d7df9088deee205ee4bc15e)
[feat: 존재하지 않는 예약 삭제시 예외 처리 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/1b416ad65ee592c2ed795e8d7e8a4e9a786abcca)

EastHShin replied:
더 많은 검증 사항이 생긴 경우는 도메인 규칙에 따른 검증 사항이라 생각되는데요~ 이 부분은 서비스 계층에서 처리해주는게 좋을 것 같아요.
다만 제가 이 코멘트를 드린 이유는 DB에서 삭제 처리가 되었는지 확인하는 부분은 도메인 규칙과 관련없이 행해져야 하는 부분이기 때문에 서비스 계층에서 이를 처리해줄 필요가 없다고 생각했기 때문이에요 :)

EastHShin commented on src/test/java/roomescape/RoomescapeApplicationTest.java:
```diff
@@ -3,6 +3,13 @@
 import org.junit.jupiter.api.Test;
 import org.springframework.boot.test.context.SpringBootTest;
 
+// TODO : 오류 나는 이유 분석해보기
+/***
+ * @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT) 형식으로 작성하면 오류가 안 남
+ *
+ */
+
+/*
 @SpringBootTest
 class RoomescapeApplicationTest {
```
분석은 끝나셨는지 궁금하네요! 👀

dye0p replied:
사실 시간이 부족하여 깊게 알아보지 못했습니다..ㅠ 우선 지금까지 알게된 부분에 대해서 정리해봤습니다!

`@SpringBootTest()`의 기본 속성은
`@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)` 으로 동작합니다.

즉, 실제 웹 서버를 띄우지 않고, Mock기반으로 동작합니다.

하지만 지금의 테스트는 실제 HTTP 요청을 보내는 형식입니다.
그래서 `@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)` 으로 변경을 해주었을 때 오류가 나지않고 테스트가 성공했는데, 이유는 해당 설정은 실제 등록된` port 번호(8080)`를 사용하기 때문에 오류가 나지 않았습니다.

만약 `@SpringBootTest()`의 기본속성으로 테스트가 동작하려면 `MockMvc` 환경을 구성해주어야 합니다.

이 정도의 흐름으로 정리하였는데, 조금 더 공부가 필요할 것 같습니다! 
혹시 부족한 부분이 있다면 피드백 부탁드립니다 : )

EastHShin replied:
아마 에러가 발생하는 이유는 어플리케이션 컨텍스트를 로드하면서 schema.sql이 이미 실행된 상태에서 또 다시 실행되어 발생하는 에러인 것 같아요~ 테스트 격리를 진행해주면 될 것 같습니다. 현재는 RoomescapeApplicationTest 를 주석처리 하더라도 동일한 이유로 다른 테스트가 깨지고 있네요 :(

EastHShin commented on src/test/java/roomescape/service/ReservationTimeServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+import roomescape.dao.FakeReservationTimeDaoImpl;
+import roomescape.dto.ReservationTimeRequestDto;
+
+public class ReservationTimeServiceTest {
```
api 레벨의 통합테스트를 한번 작성해보시는 것도 좋을 것 같아요~

dye0p replied:
Controller 테스트를 추가하였습니다! 피드백에서 의도하신 테스트가 맞을까요?!

[test: ReservationTimeController API 테스트 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/13288e9ccc0698f4e9b5d1b986da0d1758a73714)
[test: ThemeController API 테스트 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/352a71500b885039fe824e512fa62e931bd6a9cc)
[test: ReservationController API 테스트 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/f4ca11aeb449aa2f20581caa68775f907ef859ae)

EastHShin reviewed (changes requested):
안녕하세요 도기~

피드백 반영 잘 해주셨네요!
질문에 대한 답변과 추가 코멘트 남겨 두었으니 확인 부탁드려요 :)

dye0p committed:
refactor: 인기테마 조회시 최대 개수를 외부에서 받을 수 있도록 수정

dye0p committed:
feat: 커스텀 예외 이외의 예외를 처리하는 예외 핸들러 추가

- dto의 validation 예외를 처리하는 핸들러 추가
- 커스텀 예외 이외의 예외를 처리하는 핸들러 추가

dye0p committed:
refactor: 테마 삭제 유효성 검사 로직 repository 계층으로 이동

- 기존 서비스 로직에서 수행하던 삭제할 테마에 대한 유효성 검증을 Repository 계층으로 이동
- 데이터에 대한 CRDU는 비즈니스적 로직 보단, DB의 책임이라고 판단함.
- 따라서 발생해야 하는 예외 검증도 repository 계층에서 하는 것이 적절한 책임이라고 판단하여 변경하였음.

dye0p committed:
test: ReservationTimeController API 테스트 추가

dye0p committed:
test: ThemeController API 테스트 추가

dye0p committed:
test: ReservationController API 테스트 추가

dye0p committed:
test: FakeThemeDaoImpl delete 메서드 구현 추가하여 테스트 성공하도록 수정

dye0p committed:
refactor: 예약 시간 삭제시 검증 로직 repository 계층으로 이동

dye0p committed:
feat: 존재하지 않는 예약 삭제시 예외 처리 추가

dye0p committed:
test: Controller 테스트의 삭제 테스트 검증 추가

dye0p committed:
refactor: 예약 삭제 반환 예외 타입 수정

- InvalidThemeException -> InvalidReservationException

dye0p reviewed (commented):
안녕하세요 이스트! 피드백 반영과 함께 저의 생각을 코멘트로 남겼습니다! 
리뷰 잘 부탁드려요! : )

---

### 리팩터링 커밋 사항

[refactor: 인기테마 조회시 최대 개수를 외부에서 받을 수 있도록 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/5e5045ec2605c6c180d1172c8844112f81eacd1a)

[feat: 커스텀 예외 이외의 예외를 처리하는 예외 핸들러 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/64dc2abed8dbd80dd6efd58a79dea0116f600272) 

[refactor: 테마 삭제 유효성 검사 로직 repository 계층으로 이동](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/53ca3cdc0207ffd6a1c1f20b0293311c1aaff51d) 

[test: ReservationTimeController API 테스트 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/13288e9ccc0698f4e9b5d1b986da0d1758a73714)

[test: ThemeController API 테스트 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/352a71500b885039fe824e512fa62e931bd6a9cc)

[test: ReservationController API 테스트 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/f4ca11aeb449aa2f20581caa68775f907ef859ae)

[test: FakeThemeDaoImpl delete 메서드 구현 추가하여 테스트 성공하도록 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/955fdc2ccf63b8a377749a4ff1f368efd02d6222)

[refactor: 예약 시간 삭제시 검증 로직 repository 계층으로 이동](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/274f947dccce248a8d7df9088deee205ee4bc15e)

[feat: 존재하지 않는 예약 삭제시 예외 처리 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/1b416ad65ee592c2ed795e8d7e8a4e9a786abcca)

[test: Controller 테스트의 삭제 테스트 검증 추가](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/7d363ecb8b6e7f630b7bcc2679c08deb61156ae5)

[refactor: 예약 삭제 반환 예외 타입 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/3bc7c5f6c94b99b5f389fd1aacbc5825e9cb2599)

EastHShin commented on src/test/java/roomescape/controller/ReservationControllerTest.java:
```diff
@@ -0,0 +1,246 @@
+        RestAssured.port = port;
+    }
+
+    @DisplayName("예약을 등록하면 응답 예약이 반환된다.")
+    @Test
+    void createReservation() {
```
요 테스트는 실패하는 테스트네요~ 다시 한번 확인해주세요~

EastHShin reviewed (changes requested):
안녕하세요 도기~

피드백 반영 잘 해주셨네요!
테스트 깨지는 부분이 있어서 이 부분만 해소해주시면 approve and merge 하도록 하겠습니다~

dye0p committed:
test: 현재 날짜의 영향을 받는 테스트 수정

- 현재 날짜를 기준으로 프로덕션 코드가 동작했기 때문에 현재 날짜 이전의 테스트 코드가 실패하던 문제를 LocalDate.now.plusDays() 메서드를 사용하여 항상 현재 날짜 다음 날짜를 기준으로 테스트 하도록 수정

dye0p commented:
안녕하세요 이스트! 날짜값을 메서드 내부에서 정의하고 있어서 테스트의 날짜를 제어할 수 없었네요! 테스트에서 임의의 날짜를 넣어주는 것을 plusDays() 메서드를 사용하여 항상 현재 날짜 이후의 날짜로 테스트하도록 변경하였습니다!

[test: 현재 날짜의 영향을 받는 테스트 수정](https://github.com/woowacourse/spring-roomescape-member/pull/240/commits/2da7ba0d10b7ff60f1896cd7cd2160c14b6d4af3)

EastHShin reviewed (approved):
안녕하세요 도기~

미션 요구사항은 모두 충족된 듯 하여 마무리하도록 하겠습니다~
다음 단계에서 뵈어요 :)
